
# HG changeset patch
# User kcr
# Date 1455722719 28800
# Node ID d5312b49f9ad0714f8c4bc5d1614e4a2229e52d8
# Parent  7787d808c20fa71d41e7af15713bcaa6e72a4e07
8149967: Cannot compile JFXPanel with JDK 9: SurfaceData::getDefaultScale not found
Reviewed-by: flar

diff -r 7787d808c20f -r d5312b49f9ad modules/swing/src/main/java/javafx/embed/swing/JFXPanel.java
--- a/modules/swing/src/main/java/javafx/embed/swing/JFXPanel.java	Wed Feb 17 15:20:42 2016 +1300
+++ b/modules/swing/src/main/java/javafx/embed/swing/JFXPanel.java	Wed Feb 17 07:25:19 2016 -0800
@@ -79,6 +79,7 @@
 import sun.awt.CausedFocusEvent;
 import sun.awt.SunToolkit;
 import sun.java2d.SunGraphics2D;
+import sun.java2d.SurfaceData;
 import sun.util.logging.PlatformLogger;
 import sun.util.logging.PlatformLogger.Level;
 
@@ -648,6 +649,23 @@
                 e.getCaret().getInsertionIndex());
     }
 
+    // FIXME: once we move to JDK 9 as the boot JDK we should remove the
+    // reflection code from this method, consider changing it to
+    // use double rather than int, and account for the possibility of
+    // a different scale factor in X and Y.
+    private int getDefaultScale(SurfaceData surfaceData) {
+        /*
+        double scale = surfaceData.getDefaultScaleX();
+        */
+        double scale = 1;
+        try {
+            Method meth = SurfaceData.class.getMethod("getDefaultScaleX");
+            scale = (Double)meth.invoke(surfaceData);
+        } catch (Exception ex) {
+        }
+
+        return (int)Math.round(scale);
+    }
 
     /**
      * Overrides the {@link javax.swing.JComponent#paintComponent(Graphics)}
@@ -689,7 +707,7 @@
 
             int newScaleFactor = scaleFactor;
             if (g instanceof SunGraphics2D) {
-                newScaleFactor = ((SunGraphics2D)g).surfaceData.getDefaultScale();
+                newScaleFactor = getDefaultScale(((SunGraphics2D)g).surfaceData);
             }
             if (scaleFactor != newScaleFactor) {
                 createResizePixelBuffer(newScaleFactor);

